{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#habenula\n",
    "rm(list=ls())    #clear workspace\n",
    "\n",
    "library(devtools)\n",
    "library(dplyr)\n",
    "library(useful)\n",
    "library(ggplot2)\n",
    "#install_github(\"satijalab/seurat\")\n",
    "library(Seurat)\n",
    "library(stringr)\n",
    "\n",
    "#load habenula data\n",
    "setwd('/Volumes/neurobio/MICROSCOPE/Daniel/inDrops/data/habenula')\n",
    "data = read.table('habData.csv', header = T, fill = T, row.names = 1, sep =',')\n",
    "\n",
    "#fill out\n",
    "\n",
    "###########################################\n",
    "#load in to seurat\n",
    "\n",
    "hab = new('seurat', raw.data = data)\n",
    "\n",
    "#500 min genes for cortical data, 300 min genes for hab\n",
    "hab = Setup(hab, min.cells = 3, min.genes  = 300, do.logNormalize = T, total.expr = 10000, project = 'hab')\n",
    "\n",
    "# We calculate the percentage of mitochondrial genes here and store it in percent.mito using the AddMetaData.\n",
    "#The % of UMI mapping to MT-genes is a common scRNA-seq QC metric.\n",
    "\n",
    "mito.genes = grep(\"^mt.\", rownames(hab@data), value = T)\n",
    "percent.mito = colSums(expm1(hab@data[mito.genes, ]))/colSums(expm1(hab@data))\n",
    "\n",
    "#AddMetaData adds columns to object@data.info, and is a great place to stash QC stats\n",
    "hab = AddMetaData(hab, percent.mito, \"percent.mito\")\n",
    "\n",
    "VlnPlot(hab, c(\"nGene\", \"nUMI\", \"percent.mito\"), nCol = 3)\n",
    "\n",
    "#GenePlot is typically used to visualize gene-gene relationships, but can be used for anything calculated by the object,\n",
    "#i.e. columns in object@data.info, PC scores etc.\n",
    "par(mfrow = c(1, 2))\n",
    "GenePlot(hab, \"nUMI\", \"percent.mito\")\n",
    "GenePlot(hab, \"nUMI\", \"nGene\")\n",
    "\n",
    "# cutoff cells with 50% of genes coming from mitochondrial transcripts.\n",
    "hab = SubsetData(hab, subset.name = \"percent.mito\", accept.high = 0.5)\n",
    "\n",
    "#how many cells did we just cutout?\n",
    "length(hab@data[1,])\n",
    "\n",
    "#how many cells are out from right v left?\n",
    "length(grep('R', colnames(hab@data)))\n",
    "length(grep('L', colnames(hab@data)))\n",
    "#700 more from R\n",
    "\n",
    "#RegressOut regresses out unimportant effects of gene variability...Look at raw code\n",
    "hab = RegressOut(hab,latent.vars = c('nUMI','percent.mito'))\n",
    "\n",
    "#detect variable geens\n",
    "# I try to get roughly ~400 genes\n",
    "hab = MeanVarPlot(hab ,fxn.x = expMean, fxn.y = logVarDivMean, x.low.cutoff = 0.0125, x.high.cutoff = 3, y.cutoff = 2.25, do.contour = F, num.bin = 100, do.text = F)\n",
    "\n",
    "hab = PCA(hab, pc.genes = hab@var.genes, do.print = TRUE, pcs.print = 5, genes.print = 5)\n",
    "\n",
    "#ProjectPCA scores each gene in the dataset (including genes not included in the PCA) based on their correlation with the calculated components.\n",
    "#Though we donât use this further here, it can be used to identify markers that are strongly correlated with cellular heterogeneity, \n",
    "#but may not have passed through variable gene selection. The results of the projected PCA can be explored by setting use.full=T in the functions below.\n",
    "hab = ProjectPCA(hab)\n",
    "\n",
    "# Examine  and visualize PCA results a few different ways\n",
    "PrintPCA(hab, pcs.print = 1:5, genes.print = 5, use.full = TRUE)\n",
    "\n",
    "VizPCA(hab, 1:4)\n",
    "PCAPlot(hab, 1,2)\n",
    "PCHeatmap(hab, pc.use = 1:12, cells.use = 500, do.balanced = TRUE, label.columns = FALSE, use.full = FALSE)\n",
    "\n",
    "# NOTE: This process can take a long time for big datasets, comment out for expediency.\n",
    "# More approximate techniques such as those implemented in PCElbowPlot() can be used to reduce computation time\n",
    "PCElbowPlot(hab)\n",
    "\n",
    "\n",
    "hab = JackStraw(hab, num.replicate = 100, do.print = T) \n",
    "JackStrawPlot(hab,PCs=1:30)\n",
    "#looks like 19 PCs are significant, use 20\n",
    "\n",
    "#save.SNN=T saves the SNN so that the  SLM algorithm can be rerun using the same graph, but with a different resolution value (see docs for full details)\n",
    "hab = FindClusters(hab, pc.use = 1:20, resolution = 1.2, print.output = T, save.SNN = T)\n",
    "\n",
    "#runTSNE\n",
    "hab = RunTSNE(hab, dims.use = 1:20, do.fast = F)\n",
    "\n",
    "#plot TSNE\n",
    "TSNEPlot(hab, do.label = T)\n",
    "\n",
    "save(hab, file = file.path('/Volumes/neurobio/MICROSCOPE/Daniel/inDrops/data/habenula', 'habClustering_14dim_1.2res.Robj'))\n",
    "\n",
    "FeaturePlot(hab, c(\"nUMI\"),cols.use = c(\"grey\",\"blue\"))\n",
    "\n",
    "VlnPlot(hab,c('Snap25','Syn1','Slc32a1', 'Slc17a6'), use.raw = F, y.log = F)\n",
    "VlnPlot(hab,c('Esr1', 'Gal','Crh','Dbh','Th','Pcp2'), use.raw = F, y.log = F)\n",
    "VlnPlot(hab,c('Olig1','Pecam1','C1qb','C1qa', 'Aldh1l1','Aldoc'), use.raw = F, y.log = F)\n",
    "VlnPlot(hab,c('Slc17a7','Slc17a6','Slc17a8','Slc32a1', 'Gad1','Gad2'), use.raw = F, y.log = F)\n",
    "\n",
    "#find cells that are from Left\n",
    "isLeft = grepl('L',hab@cell.names, ignore.case=TRUE)\n",
    "dataL =data.frame(isLeft, row.names = hab@cell.names)\n",
    "\n",
    "hab = AddMetaData(hab, dataL, \"isLeft\")\n",
    "\n",
    "\n",
    "numCells = c()\n",
    "percentLeft = c()\n",
    "\n",
    "for( i in 0:max(as.integer(hab@data.info$res.1.2))) {\n",
    "  numCells[i+1] = (sum(hab@data.info$res.1.2==i))\n",
    "  percentLeft[i+1] = sum(as.integer(hab@data.info$res.1.2==i)*as.integer(hab@data.info$isLeft))/numCells[i+1]\n",
    "}\n",
    "\n",
    "# subset data that looks to be in neuronal clusters (and one olig cluster [4] which has nuerons): 0, 4, 5, 6, 7, 8, 11, 14, 15, 22\n",
    "hab2 = SubsetData(hab, ident.use = c('0','4','5','6','7','8','11','14','15','22'))\n",
    "\n",
    "#detect variable geens\n",
    "# I try to get roughly ~400 genes\n",
    "hab2 = MeanVarPlot(hab2 ,fxn.x = expMean, fxn.y = logVarDivMean, x.low.cutoff = 0.0125, x.high.cutoff = 3, y.cutoff = 2.25, do.contour = F, num.bin = 100, do.text = F)\n",
    "\n",
    "hab2 = PCA(hab2, pc.genes = hab2@var.genes, do.print = TRUE, pcs.print = 5, genes.print = 5)\n",
    "\n",
    "#ProjectPCA scores each gene in the dataset (including genes not included in the PCA) based on their correlation with the calculated components.\n",
    "#Though we donât use this further here, it can be used to identify markers that are strongly correlated with cellular heterogeneity, \n",
    "#but may not have passed through variable gene selection. The results of the projected PCA can be explored by setting use.full=T in the functions below.\n",
    "hab2 = ProjectPCA(hab2)\n",
    "\n",
    "# Examine  and visualize PCA results a few different ways\n",
    "PrintPCA(hab2, pcs.print = 1:5, genes.print = 5, use.full = TRUE)\n",
    "\n",
    "VizPCA(hab2, 1:4)\n",
    "PCAPlot(hab2, 1,2)\n",
    "PCHeatmap(hab2, pc.use = 1:12, cells.use = 500, do.balanced = TRUE, label.columns = FALSE, use.full = FALSE)\n",
    "\n",
    "# NOTE: This process can take a long time for big datasets, comment out for expediency.\n",
    "# More approximate techniques such as those implemented in PCElbowPlot() can be used to reduce computation time\n",
    "PCElbowPlot(hab2)\n",
    "\n",
    "\n",
    "hab2 = JackStraw(hab2, num.replicate = 100, do.print = T) \n",
    "JackStrawPlot(hab2,PCs=1:30)\n",
    "#looks like 8 PCs are significant, arbitrarily choose 15\n",
    "\n",
    "#save.SNN=T saves the SNN so that the  SLM algorithm can be rerun using the same graph, but with a different resolution value (see docs for full details)\n",
    "hab2 = FindClusters(hab2, pc.use = 1:15, resolution = 1.2, print.output = T, save.SNN = T)\n",
    "\n",
    "#runTSNE\n",
    "hab2 = RunTSNE(hab2, dims.use = 1:15, do.fast = F)\n",
    "\n",
    "#plot TSNE\n",
    "TSNEPlot(hab2, do.label = T)\n",
    "\n",
    "VlnPlot(hab2,c('Snap25','Syn1','Slc32a1', 'Slc17a6'), use.raw = F, y.log = F)\n",
    "VlnPlot(hab,c('Esr1', 'Gal','Crh','Dbh','Th','Pcp2'), use.raw = F, y.log = F)\n",
    "VlnPlot(hab2,c('Olig1','Pecam1','Mbp','C1qa', 'Aldh1l1','Aldoc'), use.raw = F, y.log = F)\n",
    "VlnPlot(hab2,c('Slc17a7','Slc17a6','Slc17a8','Slc32a1', 'Gad1','Gad2'), use.raw = F, y.log = F)\n",
    "VlnPlot(hab2,c('Pvalb','Sst','Vip','Htr2c', 'Cck','Calb2'), use.raw = F, y.log = F)\n",
    "\n",
    "VlnPlot(hab2,c('Chat','Slc18a3','Slc5a7','Slc18a2'), use.raw = F, y.log = F)\n",
    "\n",
    "\n",
    "numCells2 = c()\n",
    "percentLeft2 = c()\n",
    "\n",
    "for( i in 0:max(as.integer(hab2@data.info$res.1.2))) {\n",
    "  numCells2[i+1] = (sum(hab2@data.info$res.1.2==i))\n",
    "  percentLeft2[i+1] = sum(as.integer(hab2@data.info$res.1.2==i)*as.integer(hab2@data.info$isLeft))/numCells2[i+1]\n",
    "}\n",
    "\n",
    "save(hab2, file = file.path('/Volumes/neurobio/MICROSCOPE/Daniel/inDrops/data/habenula', 'hab2Clustering_15dim_1.2res_neuronEnriched.Robj'))\n",
    "\n",
    "savePath = '/Volumes/neurobio/MICROSCOPE/Daniel/inDrops/data/habenula'\n",
    "hab2.markers = FindAllMarkers(hab2, only.pos = T, min.pct = 0.1, thresh.use = 0.25)\n",
    "write.csv(hab2.markers,file = file.path(savePath, 'hab2_neuronEnriched_unique_geneMarkers.csv'))\n",
    "\n",
    "VlnPlot(hab2,c('Lmo3','Penk','Syt15','Wif1', 'Spata7','Th'), use.raw = F, y.log = F)\n",
    "VlnPlot(hab2,c('Gabra1','Gabrb1','Cacng2','Cacna2d1', 'C1ql3','Pcdh10'), use.raw = F, y.log = F)\n",
    "VlnPlot(hab2,c('Cacna2d1', 'Cacna2d2','Cacna2d3'), use.raw = F, y.log = F)\n",
    "VlnPlot(hab2,c('Fos', 'Fosb','Jun', 'Egr1','Nr4a2','Junb'), use.raw = F, y.log = F)\n",
    "VlnPlot(hab2,c('Syn2', 'Slc6a1','Sez6l', 'Kalrn','Gria3'), use.raw = F, y.log = F)\n",
    "\n",
    "#12/16/16\n",
    "setwd('/Volumes/neurobio/MICROSCOPE/Daniel/inDrops/data/habenula')\n",
    "load(\"hab2Clustering_15dim_1.2res_neuronEnriched.Robj\")\n",
    "\n",
    "#look for medial vs lateral habenula markers\n",
    "VlnPlot(hab2,c('Tac1', 'Oprm1','Chat', 'Il18','Drd2','Htr2c','Pcdh10','Gpr151', 'Slc17a7','Slc17a6'), use.raw = F, y.log = F)\n",
    "VlnPlot(hab2,c('Tac1', 'Chat', 'Htr2c','Pcdh10','Gpr151', 'Slc17a7','Slc17a6', 'Oprm1', 'Drd2'), use.raw = F, y.log = F)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.3.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
